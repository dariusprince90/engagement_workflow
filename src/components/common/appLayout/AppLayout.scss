.app-layout {
  height: 100vh;

  // flex container properties
  display: flex;
  flex-direction: column;

  .header-container {
    // flex item properties
    flex-grow: 0;
  }

  .body-container {
    overflow: hidden;

    // flex item properties
    flex-grow: 1;

    // flex container properties
    display: flex;
    flex-flow: row nowrap;
    justify-content: flex-start;
    align-items: stretch;

    .nav-container {
      // flex item properties
      flex-grow: 0;
    }

    .app-content-container {
      overflow: auto;

      // flex item properties
      flex-grow: 1;
    }
  }
}
