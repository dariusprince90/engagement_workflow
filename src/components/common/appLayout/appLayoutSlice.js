import { createSlice } from '@reduxjs/toolkit';

import LOCAL_STORAGE_KEYS from '../../../helpers/enums/localStorageKeys';
import * as config from '../../../configs/config';

// **********************************************************************
// * constants

// **********************************************************************
// * functions

// **********************************************************************
// * thunks

// **********************************************************************
// * initial state

const initialState = {
  applicationFaultInfo: {
    hasFault: false,
    faultMessage: null,
    faultError: null
  },

  showSideBar: false,

  metadata: {
    newAppVersionAvailable: false
  }
};

// **********************************************************************
// * slice

const appLayoutSlice = createSlice({
  name: 'appLayout',
  initialState,
  reducers: {
    applicationFaulted: (state, action) => {
      const { faultMessage, error } = action.payload;

      state.applicationFaultInfo = {
        hasFault: true,
        faultMessage: faultMessage,
        faultError: error
      };
    },

    appVersionChecked: (state) => {
      const appVersion = config.App.version;
      const localStorageKey = LOCAL_STORAGE_KEYS.app.version.name;
      const localStorageAppVersion = localStorage.getItem(localStorageKey);

      // if we have no app version in local storage
      // write the version to storage
      if (!localStorageAppVersion) {
        localStorage.setItem(localStorageKey, appVersion);
        return;
      }

      // if the app versions don't match
      // update new app version available
      if (appVersion !== localStorageAppVersion) {
        state.metadata.newAppVersionAvailable = true;
      }
    },

    hideSideBar: (state) => {
      state.showSideBar = false;
    },

    showSideBar: (state) => {
      state.showSideBar = true;
    }
  }
});

// **********************************************************************
// * actions

export const { applicationFaulted, appVersionChecked, hideSideBar, showSideBar } = appLayoutSlice.actions;

// **********************************************************************
// * selectors

export const selectApplicationFaultInfo = (state) => state.appLayout.applicationFaultInfo;
export const selectMetadata = (state) => state.appLayout.metadata;
export const selectShowSideBar = (state) => state.appLayout.showSideBar;

// **********************************************************************
// * reducer

export default appLayoutSlice.reducer;
