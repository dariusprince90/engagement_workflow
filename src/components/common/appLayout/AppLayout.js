import { useSelector } from 'react-redux';

import './AppLayout.scss';

import { selectShowSideBar } from './appLayoutSlice';
import Header from '../header/Header';
import SideBar from '../sideBar/SideBar';
import AppContent from '../appContent/AppContent';

const AppLayout = () => {
  // **********************************************************************
  // * constants

  const showSideBar = useSelector(selectShowSideBar);

  // **********************************************************************
  // * component vars

  // **********************************************************************
  // * functions

  // **********************************************************************
  // * side effects

  // **********************************************************************
  // * render

  return (
    <div className="app-layout">
      {/* header */}
      <div className="header-container">
        <Header />
      </div>

      {/* body */}
      <div className="body-container">
        {/* side bar */}
        {showSideBar && (
          <div className="nav-container">
            <SideBar />
          </div>
        )}

        {/* app content */}
        <div className="app-content-container">
          <AppContent />
        </div>
      </div>
    </div>
  );
};

export default AppLayout;
