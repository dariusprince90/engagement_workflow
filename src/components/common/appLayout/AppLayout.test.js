/* eslint-disable react/prop-types */

import ReactDOM from 'react-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import * as appLayoutSlice from './appLayoutSlice';
import AppLayout from './AppLayout';

// **********************************************************************
// * constants

const testIds = {
  header: 'header',
  appContent: 'app-content',
  sideBar: 'sideBar'
};

// **********************************************************************
// * mock external dependencies

jest.mock('react-redux', () => {
  return {
    useSelector: (callback) => {
      return callback();
    }
  };
});

jest.mock('./appLayoutSlice');

jest.mock('../header/Header', () => {
  return {
    __esModule: true,
    default: () => {
      return <fake-header data-testid={testIds.header} />;
    }
  };
});

jest.mock('../appContent/AppContent', () => {
  return {
    __esModule: true,
    default: () => {
      return <fake-app-content data-testid={testIds.appContent} />;
    }
  };
});

jest.mock('../sideBar/SideBar', () => {
  return {
    __esModule: true,
    default: () => {
      return <fake-SideBar data-testid={testIds.sideBar} />;
    }
  };
});

// **********************************************************************
// * unit tests

describe('AppLayout', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AppLayout />, div);
  });

  it('renders the correct layout', () => {
    const tree = renderer.create(<AppLayout />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the header', () => {
    render(<AppLayout />);
    expect(screen.getByTestId(testIds.header)).toBeInTheDocument();
  });

  it('renders the side bar when showSideBar is true', () => {
    appLayoutSlice.selectShowSideBar.mockReturnValue(true);
    render(<AppLayout />);
    expect(screen.getByTestId(testIds.sideBar)).toBeInTheDocument();
  });

  it('does not render the side bar when showSideBar is false', () => {
    appLayoutSlice.selectShowSideBar.mockReturnValue(false);
    render(<AppLayout />);
    expect(screen.queryByTestId(testIds.sideBar)).not.toBeInTheDocument();
  });

  it('renders the app content', () => {
    render(<AppLayout />);
    expect(screen.getByTestId(testIds.appContent)).toBeInTheDocument();
  });
});
