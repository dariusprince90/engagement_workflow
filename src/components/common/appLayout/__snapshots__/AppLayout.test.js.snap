// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLayout renders the correct layout 1`] = `
<div
  className="app-layout"
>
  <div
    className="header-container"
  >
    <fake-header
      data-testid="header"
    />
  </div>
  <div
    className="body-container"
  >
    <div
      className="app-content-container"
    >
      <fake-app-content
        data-testid="app-content"
      />
    </div>
  </div>
</div>
`;
